### PR TITLE
[SPARK-38987][SHUFFLE] Throw FetchFailedException when merged shuffle blocks are corrupted and spark.shuffle.detectCorrupt is set to true

### PR DESCRIPTION
**Pull request by akpatnam25**
### What changes were proposed in this pull request?
Adds the corruption exception handling for merged shuffle chunk when spark.shuffle.detectCorrupt is set to true(default value is true)

### Why are the changes needed?
Prior to Spark 3.0, spark.shuffle.detectCorrupt is set to true by default, and this configuration is one of the knob for early corruption detection. So the fallback can be triggered as expected.

After Spark 3.0, even though spark.shuffle.detectCorrupt is still set to true by default, but the early corruption detect knob is controlled with a new configuration spark.shuffle.detectCorrupt.useExtraMemory, and it set to false by default. Thus the default behavior, with only Magnet enabled after Spark 3.2.0(internal li-3.1.1), will disable the early corruption detection, thus no fallback will be triggered. And it will drop to throw an exception when start to read the corrupted blocks.

We handle the corrupted stream for merged blocks by throwing a FetchFailedException in this case. This will trigger a retry based on the values of spark.shuffle.detectCorrupt.useExtraMemory and spark.shuffle.detectCorrupt. 

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
- Tested on internal cluster
- Added UT

This is a PR to tackle some of the build weirdness found in PR 36601 (https://github.com/apache/spark/pull/36601). 
It contains the exact same diff. Closed that one out and recreated it here.

### CODE REVIEW

File: core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala [hunk 1]
```diff
@@ -4342,6 +4346,101 @@ class DAGSchedulerSuite extends SparkFunSuite with TempLocalSparkContext with Ti
     assertDataStructuresEmpty()
   }
 
+  test("SPARK-38987: corrupted shuffle block FetchFailure should unregister merge result") {
```
Comment (otterc) [inline core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala:4349]: Nit: rename the test to "corrupted merged shuffle block FetchFailure should unregister merge results"

File: core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala [hunk 2]
```diff
@@ -4402,12 +4501,20 @@ object DAGSchedulerSuite {
   def makeMapStatus(host: String, reduces: Int, sizes: Byte = 2, mapTaskId: Long = -1): MapStatus =
     MapStatus(makeBlockManagerId(host), Array.fill[Long](reduces)(sizes), mapTaskId)
 
-  def makeBlockManagerId(host: String): BlockManagerId = {
-    BlockManagerId(host + "-exec", host, 12345)
+  def makeBlockManagerId(host: String, isShufflePushMerger: Boolean = false): BlockManagerId = {
```
Comment (otterc) [inline core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala:4504]: how about adding execId as optional argument? Will prevent future changes to this method.

Comment (akpatnam25) [reply core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala]: agreed, changed it to accept an optional argument

**Review by otterc — APPROVED** [review]
Looks good to me

File: core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala [hunk 3]
```diff
@@ -1885,6 +1885,16 @@ private[spark] class DAGScheduler(
               mapOutputTracker.
                 unregisterMergeResult(shuffleId, reduceId, bmAddress, Option(mapIndex))
             }
+          } else {
+            // Unregister the merge result of <shuffleId, reduceId> if there is a FetchFailed event
+            // and is not a  MetaDataFetchException which is signified by bmAddress being null
+            if (bmAddress != null
+              && bmAddress.executorId.equals(BlockManagerId.SHUFFLE_MERGER_IDENTIFIER)) {
```
Comment (mridulm) [inline core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala:1892]: nit: Move the `&&` to the previous line.

File: core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala [hunk 4]
```diff
@@ -1885,6 +1885,16 @@ private[spark] class DAGScheduler(
...
+              assert(pushBasedShuffleEnabled, "Pushed based shuffle needs to " +
+                "be enabled so that merge results are present.")
```
Comment (mridulm) [inline core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala:1894]: `Pushed based shuffle` -> `Push based shuffle`

Change message to something like `Push based shuffle expected to be enabled when handling merge block fetch failure` ? To make it clear what we are expecting the condition to be for.

**Review by mridulm — COMMENTED** [review]
Just a couple of minor comments, rest looks good to me.

File: core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala (hunk 3 above)
Comment (akpatnam25) [reply core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala:1892]: done

**Review by mridulm — APPROVED** [review]
Looks good to me.

**Comment by mridulm** [comment]
Merged to master, thanks for working on this @akpatnam25 !
Thanks for the review @otterc  :-)

**Pull request closed without merging**